### PR TITLE
142682 - BESS CS Create Mock Endpoint for File Upload

### PR DESCRIPTION
**Pull request by shailen14599**
PBI consists changes of [PBI 142682](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/142682)

- Code changes to create a mock endpoint to upload BESS configuration files to azure storage
- Added functional tests
- Added steps to publish and test BESS ft project

### CODE REVIEW

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+using UKHO.BESS.API.FunctionalTests.Helpers;
+
+namespace UKHO.BESS.API.FunctionalTests.FunctionalTests
+{
+    public class UploadBessConfigFileTests
+    {
+        static readonly TestConfiguration testConfiguration = new();
+
+        [Test]
+        public async Task WhenICallUploadConfigApiWithValidConfigFile_ThenCreatedStatusCodeIsReturned()
```
Comment (shirinbt14926) [inline UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs:13]: We can rename test case either as
       ` public async Task WhenICallUploadConfigApiWithValidConfigFile_Then201StatusCodeIsReturned()`
 or
        `public async Task WhenICallUploadConfigApiWithValidConfigFile_ThenCreatedStatusIsReturned()`

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            HttpResponseMessage apiResponse = await BessUploadFileHelper.UploadConfigFile(testConfiguration.bessConfig.BaseUrl, @"./TestData/configFile.json");
+            apiResponse.StatusCode.Should().Be((HttpStatusCode)201);
+        }
+
+        [Test]
+        public async Task WhenICallUploadConfigApiWithInvalidConfigFile_ThenCreatedStatusCodeIsReturned()
```
Comment (shirinbt14926) [inline UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs:20]:
```suggestion
        public async Task WhenICallUploadConfigApiWithInvalidConfigFile_ThenBadRequestStatusCodeIsReturned()
```
as invalid config file will give 400-Bad Request

**Review by shirinbt14926 — CHANGES_REQUESTED** [review]
Few Comments to address

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs (hunk 2 above)
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Models/BESSConfigModel.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        public string? KeyFileType { get; set; }
+        public List<string>? AllowedUsers { get; set; }
+        public List<string>? AllowedUserGroups { get; set; }
+        public List<Tag>? Tags { get; set; }
+        public string? ReadMeSearchFilter { get; set; }
+        public int BatchExpiryInDays { get; set; }
+        public string? IsEnabled { get; set; }
+    }
+
+    public class Tag
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Models/BESSConfigModel.cs]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            HttpResponseMessage apiResponse = await BessUploadFileHelper.UploadConfigFile(testConfiguration.bessConfig.BaseUrl, @"./TestData/InvalidConfigFile.json");
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+namespace UKHO.BESS.API.FunctionalTests.FunctionalTests
+{
+    public class UploadBessConfigFileTests
+    {
+        static readonly TestConfiguration testConfiguration = new();
+
+        [Test]
+        public async Task WhenICallUploadConfigApiWithValidConfigFile_ThenCreatedStatusCodeIsReturned()
+        {
+            HttpResponseMessage apiResponse = await BessUploadFileHelper.UploadConfigFile(testConfiguration.bessConfig.BaseUrl, @"./TestData/configFile.json");
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/FunctionalTests/UploadBessConfigFileTests.cs]: Resolved. Thanks

File: UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Controllers/MockController.cs [hunk 6]
```diff
@@ -9,10 +11,12 @@ public class MockController : Controller
     {
         private readonly string _homeDirectoryPath;
         private readonly MockService _mockService;
+        private readonly AzureStorageService _azureStorageService;
```
[the remark this answers is not in the document]
Comment (shailen14599) [reply UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Controllers/MockController.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/TestConfiguration.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    public class TestConfiguration
+    {
+        protected IConfigurationRoot configurationRoot;
+        public BessApiConfiguration bessConfig = new();
+
+        public class BessApiConfiguration
+        {
+            public string? BaseUrl { get; set; }
+            public string? ValidConfigPath { get; set; }
+            public string? InvalidConfigPath { get; set; }
```
[the remark this answers is not in the document]
Comment (Mayuresh-Gawde) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/TestConfiguration.cs]: Resolved

**Review by shirinbt14926 — APPROVED** [review]
LGTM. Thanks @Arit14975  and @shailen14599

**Review by Mayuresh-Gawde — APPROVED** [review]
LGTM from QA!!

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by amogh14893 — APPROVED** [review]
Reviewed changes for YAML , LGTM.

**Review by DanOwenUKHO — APPROVED** [review]
LGTM

**Pull request merged**